### PR TITLE
Stop recursion in cache invalidation

### LINKED ISSUE #2143: Postgresql process segfault when running drop_chunks
**Relevant system information:**
 - OS: Centos 7.5
 - PostgreSQL version: 12.3
 - TimescaleDB version: 1.7.2
 - Installation method: YUM

**Describe the bug**
Segfault when running drop_chunks on entire database. Conditions on system:

1. Master/slave replication configured between two hosts
2. Instance has multiple databases with hypertables in each
3. `drop_chunks` is ran by crond for entire database (which sequentially means fo all hypertables in it)
4. All `drop_chunks` executed once a day at the same time to retain only last 24 hours of data
5. Data is being written all the time (even when running drop_chunks)

**To Reproduce**
No idea. It happens inconsistently

**Expected behavior**
No segfault

**Actual behavior**
Segfault

**Additional context**
Saved core dump from crash:
```
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-110.el7
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /usr/pgsql-12/bin/postgres...Reading symbols from /usr/pgsql-12/bin/postgres...(no debugging symbols found)...done.
(no debugging symbols found)...done.

warning: core file may not match specified executable file.
[New LWP 20329]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `postgres: kafkaconnect dbname 10.2.50.250(43588) BI'.
Program terminated with signal 11, Segmentation fault.
#0  0x000000000074bb4e in LockAcquireExtended ()
Missing separate debuginfos, use: debuginfo-install postgresql12-server-12.3-1PGDG.rhel7.x86_64
(gdb) where
#0  0x000000000074bb4e in LockAcquireExtended ()
#1  0x00000000007492b6 in LockRelationOid ()
#2  0x0000000000490615 in relation_open ()
#3  0x00000000004f5136 in table_open ()
#4  0x0000000000866e74 in SearchCatCacheMiss ()
#5  0x00000000008676c9 in SearchCatCache ()
#6  0x0000000000878746 in GetSysCacheOid ()
#7  0x000000000052ded5 in get_namespace_oid ()
#8  0x00007fe9af03d303 in extension_update_state () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#9  0x00007fe9af03d867 in ts_extension_is_loaded () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#10 0x00007fe9af028078 in cache_invalidate_callback () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#11 0x00000000008698ca in LocalExecuteInvalidationMessage ()
#12 0x0000000000746188 in ReceiveSharedInvalidMessages ()
#13 0x00000000007492c0 in LockRelationOid ()
#14 0x0000000000490615 in relation_open ()
#15 0x00000000004f5136 in table_open ()
#16 0x0000000000866e74 in SearchCatCacheMiss ()
#17 0x00000000008676c9 in SearchCatCache ()
#18 0x0000000000878746 in GetSysCacheOid ()
#19 0x000000000052ded5 in get_namespace_oid ()
#20 0x00007fe9af03d303 in extension_update_state () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#21 0x00007fe9af03d867 in ts_extension_is_loaded () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#22 0x00007fe9af028078 in cache_invalidate_callback () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#23 0x00000000008698ca in LocalExecuteInvalidationMessage ()
#24 0x0000000000746188 in ReceiveSharedInvalidMessages ()
#25 0x00000000007492c0 in LockRelationOid ()
#26 0x0000000000490615 in relation_open ()
#27 0x00000000004f5136 in table_open ()
#28 0x0000000000866e74 in SearchCatCacheMiss ()
#29 0x00000000008676c9 in SearchCatCache ()
#30 0x0000000000878746 in GetSysCacheOid ()
#31 0x000000000052ded5 in get_namespace_oid ()
#32 0x00007fe9af03d303 in extension_update_state () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#33 0x00007fe9af03d867 in ts_extension_is_loaded () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#34 0x00007fe9af028078 in cache_invalidate_callback () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#35 0x00000000008698ca in LocalExecuteInvalidationMessage ()
#36 0x0000000000746188 in ReceiveSharedInvalidMessages ()
#37 0x00000000007492c0 in LockRelationOid ()
#38 0x0000000000490615 in relation_open ()
#39 0x00000000004f5136 in table_open ()
#40 0x0000000000866e74 in SearchCatCacheMiss ()
#41 0x00000000008676c9 in SearchCatCache ()
#42 0x0000000000878746 in GetSysCacheOid ()
#43 0x000000000052ded5 in get_namespace_oid ()
#44 0x00007fe9af03d303 in extension_update_state () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#45 0x00007fe9af03d867 in ts_extension_is_loaded () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#46 0x00007fe9af028078 in cache_invalidate_callback () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#47 0x00000000008698ca in LocalExecuteInvalidationMessage ()
#48 0x0000000000746188 in ReceiveSharedInvalidMessages ()
#49 0x00000000007492c0 in LockRelationOid ()
#50 0x0000000000490615 in relation_open ()
#51 0x00000000004f5136 in table_open ()
#52 0x0000000000866e74 in SearchCatCacheMiss ()
#53 0x00000000008676c9 in SearchCatCache ()
#54 0x0000000000878746 in GetSysCacheOid ()
#55 0x000000000052ded5 in get_namespace_oid ()
#56 0x00007fe9af03d303 in extension_update_state () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#57 0x00007fe9af03d867 in ts_extension_is_loaded () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#58 0x00007fe9af028078 in cache_invalidate_callback () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#59 0x00000000008698ca in LocalExecuteInvalidationMessage ()
#60 0x0000000000746188 in ReceiveSharedInvalidMessages ()
#61 0x00000000007492c0 in LockRelationOid ()
---Type <return> to continue, or q <return> to quit---
#62 0x0000000000490615 in relation_open ()
#63 0x00000000004f5136 in table_open ()
#64 0x0000000000866e74 in SearchCatCacheMiss ()
#65 0x00000000008676c9 in SearchCatCache ()
#66 0x0000000000878746 in GetSysCacheOid ()
#67 0x000000000052ded5 in get_namespace_oid ()
#68 0x00007fe9af03d303 in extension_update_state () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#69 0x00007fe9af03d867 in ts_extension_is_loaded () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#70 0x00007fe9af028078 in cache_invalidate_callback () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#71 0x00000000008698ca in LocalExecuteInvalidationMessage ()
#72 0x0000000000746188 in ReceiveSharedInvalidMessages ()
#73 0x00000000007492c0 in LockRelationOid ()
#74 0x0000000000490615 in relation_open ()
#75 0x00000000004f5136 in table_open ()
#76 0x0000000000866e74 in SearchCatCacheMiss ()
#77 0x00000000008676c9 in SearchCatCache ()
#78 0x0000000000878746 in GetSysCacheOid ()
#79 0x000000000052ded5 in get_namespace_oid ()
#80 0x00007fe9af03d303 in extension_update_state () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#81 0x00007fe9af03d867 in ts_extension_is_loaded () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#82 0x00007fe9af028078 in cache_invalidate_callback () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#83 0x00000000008698ca in LocalExecuteInvalidationMessage ()
#84 0x0000000000746188 in ReceiveSharedInvalidMessages ()
#85 0x00000000007492c0 in LockRelationOid ()
#86 0x0000000000490615 in relation_open ()
#87 0x00000000004f5136 in table_open ()
#88 0x0000000000866e74 in SearchCatCacheMiss ()
#89 0x00000000008676c9 in SearchCatCache ()
#90 0x0000000000878746 in GetSysCacheOid ()
#91 0x000000000052ded5 in get_namespace_oid ()
#92 0x00007fe9af03d303 in extension_update_state () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#93 0x00007fe9af03d867 in ts_extension_is_loaded () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#94 0x00007fe9af028078 in cache_invalidate_callback () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#95 0x00000000008698ca in LocalExecuteInvalidationMessage ()
#96 0x0000000000746188 in ReceiveSharedInvalidMessages ()
#97 0x00000000007492c0 in LockRelationOid ()
#98 0x0000000000490615 in relation_open ()
#99 0x00000000004f5136 in table_open ()
#100 0x0000000000866e74 in SearchCatCacheMiss ()
#101 0x00000000008676c9 in SearchCatCache ()
#102 0x0000000000878746 in GetSysCacheOid ()
#103 0x000000000052ded5 in get_namespace_oid ()
#104 0x00007fe9af03d303 in extension_update_state () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#105 0x00007fe9af03d867 in ts_extension_is_loaded () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#106 0x00007fe9af028078 in cache_invalidate_callback () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#107 0x00000000008698ca in LocalExecuteInvalidationMessage ()
#108 0x0000000000746228 in ReceiveSharedInvalidMessages ()
#109 0x00000000007492c0 in LockRelationOid ()
#110 0x0000000000490615 in relation_open ()
#111 0x00000000004f5136 in table_open ()
#112 0x0000000000866e74 in SearchCatCacheMiss ()
#113 0x00000000008676c9 in SearchCatCache ()
#114 0x0000000000878746 in GetSysCacheOid ()
#115 0x000000000052ded5 in get_namespace_oid ()
#116 0x00007fe9af03d303 in extension_update_state () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#117 0x00007fe9af03d867 in ts_extension_is_loaded () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#118 0x00007fe9af028078 in cache_invalidate_callback () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#119 0x00000000008698ca in LocalExecuteInvalidationMessage ()
#120 0x0000000000746188 in ReceiveSharedInvalidMessages ()
#121 0x00000000007492c0 in LockRelationOid ()
#122 0x0000000000490615 in relation_open ()
#123 0x00000000004f5136 in table_open ()
---Type <return> to continue, or q <return> to quit---


AND SO ON UNTIL:

#102956 0x00007fe9af03d303 in extension_update_state () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#102957 0x00007fe9af03d867 in ts_extension_is_loaded () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#102958 0x00007fe9af028078 in cache_invalidate_callback () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#102959 0x00000000008698ca in LocalExecuteInvalidationMessage ()
#102960 0x0000000000746188 in ReceiveSharedInvalidMessages ()
#102961 0x00000000007492c0 in LockRelationOid ()
#102962 0x0000000000490615 in relation_open ()
#102963 0x00000000004f5136 in table_open ()
#102964 0x0000000000866e74 in SearchCatCacheMiss ()
#102965 0x00000000008676c9 in SearchCatCache ()
#102966 0x0000000000878746 in GetSysCacheOid ()
#102967 0x000000000052ded5 in get_namespace_oid ()
#102968 0x00007fe9af03d303 in extension_update_state () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#102969 0x00007fe9af03d71f in ts_extension_invalidate () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#102970 0x00007fe9af02806f in cache_invalidate_callback () from /usr/pgsql-12/lib/timescaledb-1.7.2.so
#102971 0x00000000008698ca in LocalExecuteInvalidationMessage ()
#102972 0x0000000000746228 in ReceiveSharedInvalidMessages ()
#102973 0x00000000007492c0 in LockRelationOid ()
#102974 0x0000000000490615 in relation_open ()
#102975 0x00000000004f5136 in table_open ()
#102976 0x0000000000866e74 in SearchCatCacheMiss ()
#102977 0x00000000008676c9 in SearchCatCache ()
#102978 0x0000000000878746 in GetSysCacheOid ()
#102979 0x000000000052ded5 in get_namespace_oid ()
---Type <return> to continue, or q <return> to quit---
#102980 0x000000000052e0ed in recomputeNamespacePath ()
#102981 0x0000000000531679 in OverrideSearchPathMatchesCurrent ()
#102982 0x000000000086df71 in RevalidateCachedQuery ()
#102983 0x000000000086e316 in GetCachedPlan ()
#102984 0x000000000076237f in PostgresMain ()
#102985 0x0000000000484022 in ServerLoop ()
#102986 0x00000000006f14c3 in PostmasterMain ()
#102987 0x0000000000484f23 in main ()

```

### PR DESCRIPTION
**Pull request by mkindahl**
Inside the `cache_invalidate_callback` the OID of the namespace is read
using `get_namespace_oid`, which usually goes to the cache. However, if
there is a cache miss inside that function, it will attempt to read the
OID from the heap, which involves invalidating the cache, which further
leads to `cache_invalidate_callback` being called recursively.

This commit prevent an infinite recursion by not doing a recursive call
and instead considering the cache invalidated. This will allow
`get_namespace_oid` (and other functions) to proceed with reading the
OID from the heap tables, fill in the cache, and return it to callers.

Closes #2143

### CODE REVIEW

**Review by pmwkaa — APPROVED** [review]
I guess it would be difficult to make a test case for it?

**Comment by mkindahl** [comment]
> I guess it would be difficult to make a test case for it?

Yes, I'm not entirely sure how the recursion is triggered. The state of the cache has to be "post" for this to trigger.

**Comment by pmwkaa** [comment]
> > I guess it would be difficult to make a test case for it?
> 
> Yes, I'm not entirely sure how the recursion is triggered. The state of the cache has to be "post" for this to trigger.

I guess we might want to initiate a call by our test C-function, not sure if the effect worth it :)

**Comment by mkindahl** [comment]
> > > I guess it would be difficult to make a test case for it?
> > 
> > 
> > Yes, I'm not entirely sure how the recursion is triggered. The state of the cache has to be "post" for this to trigger.
> 
> I guess we might want to initiate a call by our test C-function, not sure if the effect worth it :)

Hmmm.... that's an idea. I'll think about it (not sure if it really is good as a test, but maybe).

File: src/cache_invalidate.c [hunk 1]
```diff
@@ -67,6 +67,7 @@ static void
 cache_invalidate_callback(Datum arg, Oid relid)
 {
 	Catalog *catalog;
+	static bool in_recursion = false;
```
Comment (WireBaron) [inline src/cache_invalidate.c:70]: Is there a risk that multiple threads might be trying to invalidate items at the same time?

Comment (mkindahl) [reply src/cache_invalidate.c:70]: PostgreSQL is single-threaded: there is just one process that executes. This pattern is used quite often inside PostgreSQL itself, for example in `src/backend/utils/cache/inval.c`.

File: src/cache_invalidate.c [hunk 2]
```diff
@@ -77,16 +78,27 @@ cache_invalidate_callback(Datum arg, Oid relid)
...
+	/* The cache invalidation can be called indirectly further down in the
+	 * call chain by calling `get_namespace_oid`, which can trigger a
+	 * recursive cache invalidation callback. To prevent infinite recursion,
+	 * we stop it here and instead consider the cache as invalidated, which
+	 * will allow the `get_namespace_oid` to retrieve the OID from the heap
+	 * table and return it properly. */
+	if (!in_recursion)
+	{
+		in_recursion = true;
+		catalog = ts_catalog_get();
```
Comment (erimatnor) [inline src/cache_invalidate.c:90]: Nit: you can now move the `Catalog` declaration inside the local context here.

**Review by erimatnor — APPROVED** [review]
Agree on it being nice to produce a test case, but I guess we didn't have one for the previous recursive cache invalidation fix either.

File: src/cache_invalidate.c (hunk 2 above)
Comment (mkindahl) [reply src/cache_invalidate.c:90]: Moved.

**Pull request merged**